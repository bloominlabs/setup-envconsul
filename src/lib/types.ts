import * as rt from "runtypes";

const OsRt = rt.Union(
  rt.Literal("darwin"),
  rt.Literal("dragonfly"),
  rt.Literal("freebsd"),
  rt.Literal("linux"),
  rt.Literal("netbsd"),
  rt.Literal("openbsd"),
  rt.Literal("solaris"),
  rt.Literal("windows")
);

const ArchRt = rt.Union(
  rt.Literal("arm"),
  rt.Literal("arm64"),
  rt.Literal("amd64"),
  rt.Literal("386"),
  rt.Literal("mips"),
  rt.Literal("mips64"),
  rt.Literal("mipsle"),
  rt.Literal("s390x"),
  rt.Literal("ppc64le"),
  rt.Literal("amd64-lxc"),
  rt.Literal("arm5"),
  rt.Literal("arm6"),
  rt.Literal("arm7")
);

const BuildRt = rt.Record({
  name: rt.String,
  version: rt.String,
  os: OsRt,
  arch: ArchRt,
  filename: rt.String,
  url: rt.String,
});

const VersionRt = rt.Record({
  name: rt.String,
  version: rt.String,
  shasums: rt.String,
  shasums_signature: rt.String,
  shasums_signatures: rt.Array(rt.String),
  builds: rt.Array(BuildRt),
});

export const IndexRt = rt.Record({
  name: rt.String,
  versions: rt.Dictionary(VersionRt),
});

export type Index = rt.Static<typeof IndexRt>;
export type Version = rt.Static<typeof VersionRt>;
