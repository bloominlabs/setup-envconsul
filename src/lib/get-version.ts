import { maxSatisfying, gte } from "semver";
import * as types from "./types";

export async function getVersionObject(
  index: types.Index,
  range: string
): Promise<types.Version> {
  const versions = index.versions;
  if (range == "latest") {
    const latest = Object.keys(versions).reduce((prev, cur) => {
      return gte(cur, prev) ? cur : prev;
    });
    invariant(latest, "expect a latest version to exists");
    return versions[latest];
  }

  const resp = maxSatisfying(Object.keys(versions), range);
  console.log(resp);
  if (resp === null) {
    throw new Error(
      "Could not find a version that satisfied the version range"
    );
  }

  const ver = versions[resp];
  if (!ver) {
    throw new Error(
      "Could not find a version that satisfied the version range"
    );
  }

  return ver;
}

/* eslint @typescript-eslint/explicit-module-boundary-types: 0 */
export function invariant(
  condition: unknown,
  message?: string
): asserts condition {
  if (!condition) {
    throw new Error(message);
  }
}
