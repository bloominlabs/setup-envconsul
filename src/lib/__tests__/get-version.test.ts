import got from "got/dist/source";
import * as playback from "jest-playback";
import * as types from "../types";
import { getVersionObject } from "../get-version";
playback.setup(__dirname);

describe("get-version", () => {
  process.env.GITHUB_TOKEN = process.env.GITHUB_TOKEN || "my-token";
  describe("range versions - envconsul", () => {
    it.each([
      "latest",
      "^0",
      "0.*.*",
      "0.2.*",
      "v0.*.*",
      "v0.12.*",
      "0.12.1",
      "v0.12.0",
      "0.12.1",
    ] as const)("should match %s versions", async (ver) => {
      const result = await got(
        "https://releases.hashicorp.com/envconsul/index.json",
        {
          responseType: "json",
        }
      );

      const v = await getVersionObject(types.IndexRt.check(result.body), ver);
      expect(v.version).toMatchSnapshot();
    });
  });
});
