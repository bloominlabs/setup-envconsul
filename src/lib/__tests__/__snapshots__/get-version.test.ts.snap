// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`get-version range versions - envconsul should match ^0 versions 1`] = `"0.12.1"`;

exports[`get-version range versions - envconsul should match 0.*.* versions 1`] = `"0.12.1"`;

exports[`get-version range versions - envconsul should match 0.2.* versions 1`] = `"0.2.0"`;

exports[`get-version range versions - envconsul should match 0.12.1 versions 1`] = `"0.12.1"`;

exports[`get-version range versions - envconsul should match 0.12.1 versions 2`] = `"0.12.1"`;

exports[`get-version range versions - envconsul should match latest versions 1`] = `"0.12.1"`;

exports[`get-version range versions - envconsul should match v0.*.* versions 1`] = `"0.12.1"`;

exports[`get-version range versions - envconsul should match v0.12.* versions 1`] = `"0.12.1"`;

exports[`get-version range versions - envconsul should match v0.12.0 versions 1`] = `"0.12.0"`;

exports[`get-version range versions - nomad should match latest versions 1`] = `"1.3.0-beta.1+ent"`;

exports[`get-version range versions - nomad should match v1.2 versions 1`] = `"1.2.6"`;

exports[`get-version range versions - packer should match latest versions 1`] = `"1.8.0"`;
